sdcm/logcollector.py

```diff
@@ -0,0 +1,819 @@
+        local_dir = os.path.join(base_local_dir, self.current_run,
+                                 "{}-{}".format(self.cluster_log_type, self.test_id[:8]))
+        try:
+            os.makedirs(local_dir)
+        except OSError as details:
+            if not os.path.exists(local_dir):
+                LOGGER.error("Folder is not created. {}".format(details))
+                raise
+        return local_dir
+
+    def create_remote_storing_dir(self, node, path=''):
+        if not path:
+            path = node.name
+        remote_dir = os.path.join(self.node_remote_dir, path)
+        result = node.remoter.run('mkdir -p {}'.format(remote_dir), ignore_status=True)
+
+        if result.exited > 0:
+            LOGGER.error(
+                'Remote storing folder not created.\n{}'.format(result))
+
+        return remote_dir
+
+    @staticmethod
+    def collect_log_remotely(node, cmd, log_filename):
+        collect_log_command = "{cmd} >& {log_filename}".format(**locals())
+        result = node.remoter.run(
+            collect_log_command, ignore_status=True, verbose=True)
+        result = node.remoter.run(
+            'test -f {}'.format(log_filename), ignore_status=True)
+        return log_filename if result.exited == 0 else False
+
+    @staticmethod
+    def archive_log_remotely(node, log_filename, archive_name=None):
+        archive_dir = os.path.dirname(log_filename)
+        if archive_name:
+            archive_name = os.path.join(archive_dir, archive_name)
+        else:
+            archive_name = os.path.join(archive_dir, os.path.basename(log_filename))
+        log_filename = os.path.basename(log_filename)
+        result = node.remoter.run(
+            "cd {archive_dir}; tar -czf {archive_name}.tar.gz {log_filename}".format(**locals()), ignore_status=True)
+        return "{}.tar.gz".format(archive_name)
+
+    @staticmethod
+    def receive_log(node, remote_log_path, local_dir):
+        if not os.path.exists(local_dir):
+            os.mkdir(local_dir)
+        node.remoter.receive_files(src=remote_log_path, dst=local_dir)
+        return local_dir
+
+    @staticmethod
+    def upload_logs(archive_path, storing_path):
+        s3_link = S3Storage().upload_file(file_path=archive_path, dest_dir=storing_path)
+        return s3_link
+
+    def collect_logs(self, local_search_path=None):
+        if not self.nodes:
+            LOGGER.warning('No any running instances for {}'.format(self.cluster_log_type))
+            return self.local_dir, None, None
+
+        def collect_logs_per_node(node):
+            remote_node_dir = self.create_remote_storing_dir(node)
+            local_node_dir = os.path.join(self.local_dir, node.name)
+            os.makedirs(local_node_dir)
+            for log_entity in self.log_entities:
+                log_entity.collect(node, local_node_dir, remote_node_dir, local_search_path=local_search_path)
+
+        ParallelObject(self.nodes, num_workers=3, timeout=300).run(collect_logs_per_node)
```

Comment:
why num_workers is only 3 ? what if we have 10 nodes ? or more ? would it work, or just work slower ?

Reply:
will work slowly. We can use : int(num of nodes / 2) to prevent overload if, for example, there are 500 nodes?

Reply:
implemented